pkg/diagnostic/kms/check.go

```diff
@@ -23,46 +28,130 @@ var (
 )
 
 type checker struct {
-	cfg    *config.Settings
-	logger *logrus.Entry
+	cfg       *config.Settings
+	logger    *logrus.Entry
+	clientset kubernetes.Interface
 }
 
-func NewProvider(ctx context.Context, cfg *config.Settings) diagnostic.Provider {
+var NewProvider = func(ctx context.Context, cfg *config.Settings, clientset ...kubernetes.Interface) diagnostic.Provider {
+	var cs kubernetes.Interface
+	if len(clientset) > 0 {
+		cs = clientset[0]
+	} else {
+		// Use the in-cluster config if running inside a cluster, otherwise use the default kubeconfig
+		config, err := rest.InClusterConfig()
+		if err != nil {
+			kubeconfig := clientcmd.RecommendedHomeFile
+			config, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+			if err != nil {
+				panic(err.Error())
+			}
+		}
+
+		// Create the clientset
+		cs, err = kubernetes.NewForConfig(config)
+		if err != nil {
+			panic(err.Error())
+		}
+	}
+
 	return &checker{
 		cfg: cfg,
 		logger: logging.NewLogger().
 			WithContext(ctx).WithField(logging.OpField, "ksm"),
+		clientset: cs,
 	}
 }
 
-func (c *checker) Check(ctx context.Context, client *net.Client, accessor status.Accessor) error {
+func (c *checker) Check(ctx context.Context, client *http.Client, accessor status.Accessor) error {
 	var (
-		err              error
 		retriesRemaining = MaxRetry
-		url              = fmt.Sprintf("%s/", c.cfg.Prometheus.KubeStateMetricsServiceEndpoint)
+		namespace        = "prom-agent"
+		serviceName      = "cz-prom-agent-kube-state-metrics"
+		endpointURL      string
 	)
 
-	// We need to build in a retry here because the kube-state-metrics service can take a few seconds to start up
-	// If it is deploying with the cloudzero-agent chart
-	for {
-		_, err = http.Do(ctx, client, net.MethodGet, nil, nil, url, nil)
-		if err == nil {
-			break
+	// Wait for the pod to become ready and find the first available endpoint
+	for retriesRemaining > 0 {
+		endpoints, err := c.clientset.CoreV1().Endpoints(namespace).Get(ctx, serviceName, metav1.GetOptions{})
+		if err != nil {
+			c.logger.Errorf("Failed to get service endpoints: %v", err)
+			accessor.AddCheck(&status.StatusCheck{Name: DiagnosticKMS, Passing: false, Error: fmt.Sprintf("Failed to get service endpoints: %s", err.Error())})
+			return nil
 		}
-		if retriesRemaining == 0 {
+
+		// Log the endpoints for debugging
+		c.logger.Infof("Endpoints: %v", endpoints)
+
+		// Check if there are any ready addresses and find the first available endpoint
+		for _, subset := range endpoints.Subsets {
```

Comment:
I don't think I follow - why do we need to find the address and port? if we're managing KSM, we'll know what the dns and port will be, right? why not just check the given Service address?

Reply:
> The Service Endpoint provided by the Chart doesn't actually work from within the container

meaning, you can't hit `http://<release-name>.<release-namesapce>.svc.cluster.local` from within the container? that's surprising to me. that's how I generally address other pods within a cluster. for example, [here](https://github.com/Cloudzero/cloudzero-charts/blob/feature/1.0.2-beta-release/charts/cloudzero-insights-controller/templates/init-job.yaml#L28)

> There's not an easy way to have that value statically defined and passed from the chart.

isn't the KSM service name always `cloudzero-state-metrics`? because of [this](https://github.com/Cloudzero/cloudzero-charts/blob/feature/1.0.2-beta-release/charts/cloudzero-agent/values.yaml#L65-L66) override. and then we already know the namespace that the chart is deployed in. so I would think you'd be able to put the following in the validator config:

```yaml
kube_state_metrics_endpoint: https://cloudzero-state-metrics.{{ .Release.Namespace }}.svc.cluster.local:8080
```
or, you could probably get the svc name and port dynamically as well:
```yaml
kube_state_metrics_endpoint: https://{{ .Values.kubeStateMetrics.nameOverride }}.{{ .Release.Namespace }}.svc.cluster.local:{{ .Values.kubeStateMetrics.service.port }}
```

I might be missing the point

Reply:
Yea this seems right to me. What through me off with the service endpoint was the off formatting that the Agent Chart currently passes to the Validator. I tried with the formatting you mentioned above and it worked, I'll make the change. Thanks!

Reply:
yeah for sure! yeah it's a weird thing to get the name from the ksm subchart, but I'm glad it works!